### ISSUE TITLE
error: no input devices found (do you have rw permission on /dev/input/*?)

### ISSUE THREAD

**Issue by rxng** [rbreaves/kinto#553]
I am installing on popos.

After the keyboard assistant appears and I press the 2nd key, the Kinto.sh window pops up. 

However, I get this error and it seems that xkeysnail isn't picking up my USB keyboard? How do I ensure Kintos can work?

![Screenshot from 2021-07-29 11-34-07](https://user-images.githubusercontent.com/19649766/127417865-52cb339b-616e-4fa5-8033-5105ba9e3103.png)

**Comment by rbreaves** [comment]
Looks like the limitedadmins file in /etc/sudoers.d does not have the right permissions and could be part of your issue.

My script ought to run this to resolve uinput permissions, but then limitedadmins ought to allow certain commands to run elevated as well.
```
xhost +SI:localuser:root
```

How or why the permissions would be world writable I am not sure..

```
sudo chown root:root ./linux/limitedadmins.new
# Add a check here for xkeysnail path resolving
sudo mv ./linux/limitedadmins.new /etc/sudoers.d/limitedadmins
```

I don't see that I explicitly set the permissions for limitedadmins but from reading online I think you would want to do this to secure the file to be limited on who can write it. Will leave this ticket open, as I ought to revisit this and make this explicit in the installer I imagine.
```
sudo chmod 0440 /etc/sudoers.d/limitedadmins
```

**Comment by rxng** [comment]
certainly a mystery! the permissions issue was right from the start when running the command to install kinto. This output might be helpful?

```
Installing Kinto...
fatal: not a git repository (or any parent up to mount point /home/xxxx)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
fatal: not a git repository (or any parent up to mount point /home/xxxx)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
fatal: not a git repository (or any parent up to mount point /home/xxxx)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).

Kinto  build 
Type in Linux like it's a Mac.

Overlay key,  'super' , detected. Will be removing so Super-Space can remap to Cmd-Space for app launching..
Hide/minimize Window is already bound to  @as []  , please remap it to Super-H for kinto.
gsettings set org.gnome.desktop.wm.keybindings minimize "['<Super>h','<Alt>F9']"
Requirement already satisfied: pillow in /home/xxxx/.local/lib/python3.9/site-packages (8.3.1)
Using systemd...
localuser:root being added to access control list
fatal: destination path 'xkeysnail' already exists and is not an empty directory.
error: pathspec 'kinto' did not match any file(s) known to git
Processing /home/xxx/Downloads/kinto-master/xkeysnail
Requirement already satisfied: evdev in /usr/lib/python3/dist-packages (from xkeysnail==0.3.0) (1.4.0)
Requirement already satisfied: inotify_simple in /usr/local/lib/python3.9/dist-packages (from xkeysnail==0.3.0) (1.3.5)
Requirement already satisfied: python-xlib in /usr/lib/python3/dist-packages (from xkeysnail==0.3.0) (0.29)
Building wheels for collected packages: xkeysnail
  Building wheel for xkeysnail (setup.py) ... done
  Created wheel for xkeysnail: filename=xkeysnail-0.3.0-py3-none-any.whl size=17545 sha256=93ac91af2af7d8f1b6a430f09dfc5388231e02d5f2434d50376759133e7ea6d4
  Stored in directory: /tmp/pip-ephem-wheel-cache-5ct_6h5w/wheels/7e/1c/1b/c0df7ab905fdf383dc2c1b0fdb617121dd161c72b72f4675c7
Successfully built xkeysnail
Installing collected packages: xkeysnail
  Attempting uninstall: xkeysnail
    Found existing installation: xkeysnail 0.3.0
    Uninstalling xkeysnail-0.3.0:
      Successfully uninstalled xkeysnail-0.3.0
Successfully installed xkeysnail-0.3.0
/usr/local/bin/xkeysnail
sudo: /etc/sudoers.d/limitedadmins is world writable
Service file added to /usr/lib/systemd/system/xkeysnail.service
sudo: /etc/sudoers.d/limitedadmins is world writable
Ownership set for root...
sudo: /etc/sudoers.d/limitedadmins is world writable
Permissions set to 644...
sudo: /etc/sudoers.d/limitedadmins is world writable
Created soft symlink...
sudo: /etc/sudoers.d/limitedadmins is world writable
Created soft symlink for graphical target...
sudo: /etc/sudoers.d/limitedadmins is world writable
sudo: /etc/sudoers.d/limitedadmins is world writable
Removed /etc/systemd/system/graphical.target.wants/xkeysnail.service.
Removed /etc/systemd/system/xkeysnail.service.
sudo: /etc/sudoers.d/limitedadmins is world writable
Adding xhost fix...

Kinto install is complete.
```

**Comment by rxng** [comment]
@rbreaves the issue wasn't resolved by fixing limitedadmin permissions. Is there a way to specify the usb keyboard as device if it is not being picked up? See below for the output

![Screenshot from 2021-07-29 14-50-04](https://user-images.githubusercontent.com/19649766/127433598-241ab1b0-0330-455d-8592-bcd8631b7df9.png)

**Comment by rbreaves** [comment]
Well that screenshot literally says to use “--device” & lists off your Logitech Mx keyboard input devents as event2 & 3 so certainly. Just update your systemd xkeysnail.service file. There’s also plenty of past documentation here that delves into that & specifically in the case of MX keyboards.

If this also didn’t have the permissions issue on the limitedadmins I’d have closed this issue ticket already as a duplicate open issue already. I get a new issue ticket on these MX keyboards every few months it seems. They do work but yes you have to set it manually.

**Comment by rxng** [comment]
Thanks so much for your reply @rbreaves I will investigate and let you know. Been many hours on this!

**Comment by rxng** [comment]
@rbreaves yes that worked editing the xkeysnail.services file! Thank you so much!!

There is just one issue, because it seems that the super is now physical-ctrl? I saw the readme and wondered about this set up. 

Because the original mac layout is
Ctrl - Opt - Cmd

Kinto seems to make it
Cmd - Opt - Ctrl

Is there a reason for this? Why not keep it as Ctrl - Opt - Cmd? Meaning that Kinto would make it Ctrl - Alt - Super+Ctrl

Or is that not possible....

**Comment by rbreaves** [comment]
@rxng it’s a catch all type theory approach - I reorder the keys in the order in which they are used. It’s also much easier to remap Super (on physical Ctrl) & then to the few Ctrl shortcuts people use it for vs Super key on the Physical Alt key position & then the numerous Ctrl combos that would explode into.

The entire point is to do as little remapping work as possible while making it feel as it should.

This approach, I admit, presents interesting problems & issues for Remote Desktop into macOS at times whether it be from Linux or Windows. I will likely be making an update later that’ll remap Super next to the space bar by default but then still remap Ctrl to it by default as well. Reason being for better Remote Desktop compatibility. Seems crazy to do a remap twice & on 2 levels but it makes sense to do so.

If you want you can already do this - install sharpkeys & remap Super to the Cmd position & Alt to Super. Then in the Kinto setup specify your keyboard as an Apple keyboard. If your keyboard supports an Apple mode natively then that works too but the media keys might not, so sharpkeys & the Windows mode may be the preferred route still.

**Comment by rbreaves** [comment]
@rxng  Also just to be clear the Super key, while in the physical Ctrl location does behave as Ctrl when it needs to most of the time, for things such as the terminal, but yes not 100% all of the time. It is easier imo to remap Super to Ctrl when it needs to be (while in the phsyical Ctrl location) than to remap Super to Ctrl while it sits right next to the spacebar. The latter is a A LOT more work - and this project would never get the coverage you see it having right now (w/ minimal effort) and I'd be dying a death of a thousand paper cuts trying.. as I feel like everyone that has ever tried doing a project like this already has. 😂

**Comment by rxng** [comment]
@rbreaves I saw that gnome tweaks also allows you to "Ctrl is mapped to Win and other Ctrl keys", which seems to be similar.

One fix I found for logitech that avoids the sharpkeys method, is to download Solaar, and then change the keyboard type to Apple, then also set Kinto to Apple (I am on Linux). This seemed to do the trick to fix the Ctrl/Super issue locally, as well as if you are using Synergy/Barrier.
